README
NAME
    App::UniqFiles - Report or omit duplicate file contents

VERSION
    This document describes version 0.08 of App::UniqFiles (from Perl
    distribution App-UniqFiles), released on 2014-08-16.

SYNOPSIS
     # See uniq-files script

FUNCTIONS
  uniq_files(%args) -> [status, msg, result, meta]
    Report or omit duplicate file contents.

    Examples:

     uniq_files();

    List all files which do no have duplicate contents.

     uniq_files();

    List all files which have duplicate contents.

     uniq_files();

    List number of occurences of contents for each file.

    Given a list of filenames, will check each file size and content for
    duplicate content. Interface is a bit like the "uniq" Unix command-line
    program.

    Arguments ('*' denotes required arguments):

    *   check_content => *bool* (default: 1)

        Whether to check file content .

        If set to 0, uniqueness will be determined solely from file size.
        This can be quicker but might generate a false positive when two
        files of the same size are deemed as duplicate even though their
        content are different.

    *   count => *bool* (default: 0)

        Whether to return each file content's number of occurence.

        1 means the file content is only encountered once (unique), 2 means
        there is one duplicate, and so on.

    *   files* => *array*

    *   report_duplicate => *int* (default: 2)

        Whether to return duplicate items.

        Can be set to either 0, 1, 2.

        If set to 2 (the default), will only return the first of duplicate
        items. For example: file1 contains text 'a', file2 'b', file3 'a'.
        Only file1 will be returned because file2 is unique and file3
        contains 'a' (already represented by file1).

        If set to 1, will return all the the duplicate items. From the above
        example: file1 and file3 will be returned.

        If set to 0, duplicate items will not be returned.

    *   report_unique => *bool* (default: 1)

        Whether to return unique items.

    Return value:

    Returns an enveloped result (an array).

    First element (status) is an integer containing HTTP status code (200
    means OK, 4xx caller error, 5xx function error). Second element (msg) is
    a string containing error message, or 'OK' if status is 200. Third
    element (result) is optional, the actual result. Fourth element (meta)
    is called result metadata and is optional, a hash that contains extra
    information.

     (any)

NOTES
    Warning: cannot properly handle symlinks or special files (socket, pipe,
    device), so don't feed them.

TODO
    *   Handle symlinks

        Provide options on how to handle symlinks: ignore them? Follow?
        Also, with return_duplicate=2, we should not use the symlink
        (because one of the usage of uniq-files might be to delete duplicate
        files).

    *   Handle special files (socket, pipe, device)

        Ignore them.

    *   Check hardlinks/inodes first

        For fast checking.

    *   Arguments hash_skip_bytes & hash_bytes

        For only checking uniqueness against parts of contents.

    *   Arguments hash_module/hash_method/hash_sub

        For doing custom hashing instead of Digest::MD5.

HOMEPAGE
    Please visit the project's homepage at
    <https://metacpan.org/release/App-UniqFiles>.

SOURCE
    Source repository is at
    <https://github.com/sharyanto/perl-App-UniqFiles>.

BUGS
    Please report any bugs or feature requests on the bugtracker website
    <https://rt.cpan.org/Public/Dist/Display.html?Name=App-UniqFiles>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for App-UniqFiles

0.09    2015-01-03 (PERLANCAR)

	- No functional changes.

	- Rebuild (Generate TODO.md).


0.08     2014-08-16 (SHARYANTO) - Happy 19th CPAN Day!

         - [build] CLI scripts are now generated automatically using
           Dist::Zilla::Plugin::ScriptFromFunc. The plugin is still at 0.01 so the
           generated script is not as polished as it should be. Stay tuned.


0.07     2014-07-22 (SHARYANTO)

         - No functional changes.

         - Switch CLI scripts from using Perinci::CmdLine to
           Perinci::CmdLine::Any to reduce size of dependencies.


0.06    2014-05-16 (SHARYANTO)

        - No functional changes.

        - Replace File::Slurp with File::Slurp::Tiny.


0.05    2012-03-02 (SHARYANTO)

        - No functional changes.

        - Replace Sub::Spec with Rinci/Perinci.


0.04    2011-07-28 (SHARYANTO)

        [ENHANCEMENTS]

        - Add option: check_content.


0.03    2011-07-28 (SHARYANTO)

        - Fix behaviour to match `uniq` command: by default, duplicate files are
          returned but only one for each content.


0.02    2011-07-28 (SHARYANTO)

        - Minor bug fixes; use logging in script.


0.01    2011-07-28 (SHARYANTO)

        - First release.


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Report or omit duplicate file contents",
  "dist_author" => [
    "Steven Haryanto <stevenharyanto\@gmail.com>"
  ],
  "dist_name" => "App-UniqFiles",
  "dist_version" => "0.05",
  "license" => "perl",
  "module_name" => "App::UniqFiles",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Digest::MD5" => 0,
    "File::Slurp" => 0,
    "File::chdir" => 0,
    "Log::Any" => 0,
    "Log::Any::App" => 0,
    "Perinci::CmdLine" => 0,
    "Rinci" => "1.1.0",
    "Test::More" => "0.96",
    "perl" => "5.010000"
  },
  "script_files" => [
    "bin/uniq-files"
  ]
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.029.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Report or omit duplicate file contents",
  "AUTHOR" => "perlancar <perlancar\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Perl::osnames" => "0.09"
  },
  "DISTNAME" => "App-UniqFiles",
  "EXE_FILES" => [
    "bin/uniq-files"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "App::UniqFiles",
  "PREREQ_PM" => {
    "File::chdir" => 0,
    "Log::Any" => 0,
    "Perinci::CmdLine::Any" => "0.06",
    "Rinci" => "1.1.0"
  },
  "TEST_REQUIRES" => {
    "File::Slurp::Tiny" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.98"
  },
  "VERSION" => "0.09",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Slurp::Tiny" => 0,
  "File::Spec" => 0,
  "File::chdir" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Log::Any" => 0,
  "Perinci::CmdLine::Any" => "0.06",
  "Perl::osnames" => "0.09",
  "Rinci" => "1.1.0",
  "Test::More" => "0.98"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);




# modify generated Makefile to generate shell completion scripts. this piece
# is generated by Dist::Zilla::Plugin::GenShellCompletion version 0.07.
GEN_SHELL_COMPLETION:
{
    use Perl::osnames 0.09 qw(is_posix);
    last unless is_posix();

    print "Modifying Makefile to generate shell completion on install\n";
    open my($fh), "<", "Makefile" or die "Can't open generated Makefile: $!";
    my $content = do { local $/; ~~<$fh> };

    $content =~ s/^(install :: pure_install doc_install)/$1 comp_install/m
        or die "Can't find pattern in Makefile (1)";

    $content =~ s/^(uninstall :: .+)/$1 comp_uninstall/m
        or die "Can't find pattern in Makefile (2)";

    $content .= qq|\ncomp_install :\n\t| .
        q|$(PERLRUN) -E'if(eval { require App::shcompgen; 1 }) { system "shcompgen", "--verbose", "generate", "--replace", @ARGV }' -- $(EXE_FILES)| .
        qq|\n\n|;

    $content .= qq|\ncomp_uninstall :\n\t| .
        q|$(PERLRUN) -E'if(eval { require App::shcompgen; 1 }) { system "shcompgen", "--verbose", "remove", @ARGV }' -- $(EXE_FILES)| .
        qq|\n\n|;

    open $fh, ">", "Makefile" or die "Can't write modified Makefile: $!";
    print $fh $content;
}


lib/App/UniqFiles.pm
package App::UniqFiles;

our $DATE = '2015-01-03'; # DATE
our $VERSION = '0.09'; # VERSION

use 5.010001;
use strict;
use warnings;
use Log::Any qw($log);

use Digest::MD5;

require Exporter;
our @ISA       = qw(Exporter);
our @EXPORT_OK = qw(uniq_files);

our %SPEC;

$SPEC{uniq_files} = {
    v => 1.1,
    summary => 'Report or omit duplicate file contents',
    description => <<'_',

Given a list of filenames, will check each file size and content for duplicate
content. Interface is a bit like the `uniq` Unix command-line program.

_
    args    => {
        files => {
            schema => ['array*' => {of=>'str*'}],
            req    => 1,
            pos    => 0,
            greedy => 1,
        },
        report_unique => {
            schema => [bool => {default=>1}],
            summary => 'Whether to return unique items',
            cmdline_aliases => {
                u => {
                    summary => 'Alias for --report-unique --report-duplicate=0',
                    code => sub {
                        my $args = shift;
                        $args->{report_unique}    = 1;
                        $args->{report_duplicate} = 0;
                    },
                },
                d => {
                    summary =>
                        'Alias for --noreport-unique --report-duplicate=1',
                    code => sub {
                        my $args = shift;
                        $args->{report_unique}    = 0;
                        $args->{report_duplicate} = 1;
                    },
                },
            },
        },
        report_duplicate => {
            schema => [int => {default=>2}],
            summary => 'Whether to return duplicate items',
            description => <<'_',

Can be set to either 0, 1, 2.

If set to 2 (the default), will only return the first of duplicate items. For
example: file1 contains text 'a', file2 'b', file3 'a'. Only file1 will be
returned because file2 is unique and file3 contains 'a' (already represented by
file1).

If set to 1, will return all the the duplicate items. From the above example:
file1 and file3 will be returned.

If set to 0, duplicate items will not be returned.

_
            cmdline_aliases => {
            },
        },
        check_content => {
            schema => [bool => {default=>1}],
            summary => "Whether to check file content ",
            description => <<'_',

If set to 0, uniqueness will be determined solely from file size. This can be
quicker but might generate a false positive when two files of the same size are
deemed as duplicate even though their content are different.

_
        },
        count => {
            schema => [bool => {default=>0}],
            summary => "Whether to return each file content's ".
                "number of occurence",
            description => <<'_',

1 means the file content is only encountered once (unique), 2 means there is one
duplicate, and so on.

_
        },
    },
    examples => [
        {
            summary   => 'List all files which do no have duplicate contents',
            src       => 'uniq-files *',
            src_plang => 'bash',
        },
        {
            summary   => 'List all files which have duplicate contents',
            src       => 'uniq-files -d *',
            src_plang => 'bash',
        },
        {
            summary   => 'List number of occurences of contents for each file',
            src       => 'uniq-files -c *',
            src_plang => 'bash',
        },
    ],
};
sub uniq_files {
    my %args = @_;

    my $files = $args{files};
    return [400, "Please specify files"] if !$files || !@$files;
    my $report_unique    = $args{report_unique}    // 1;
    my $report_duplicate = $args{report_duplicate} // 2;
    my $check_content    = $args{check_content}    // 1;
    my $count            = $args{count}            // 0;

    # get sizes of all files
    my %size_counts; # key = size, value = number of files having that size
    my %file_sizes; # key = filename, value = file size, for caching stat()
    for my $f (@$files) {
        my @st = stat $f;
        unless (@st) {
            $log->error("Can't stat file `$f`: $!, skipped");
            next;
        }
        $size_counts{$st[7]}++;
        $file_sizes{$f} = $st[7];
    }

    # calculate digest for all files having non-unique sizes
    my %digest_counts; # key = digest, value = num of files having that digest
    my %digest_files; # key = digest, value = [file, ...]
    my %file_digests; # key = filename, value = file digest
    for my $f (@$files) {
        next unless defined $file_sizes{$f};
        next if $size_counts{ $file_sizes{$f} } == 1;
        my $digest;
        if ($check_content) {
            my $fh;
            unless (open $fh, "<", $f) {
                $log->error("Can't open file `$f`: $!, skipped");
                next;
            }
            my $ctx = Digest::MD5->new;
            $ctx->addfile($fh);
            $digest = $ctx->hexdigest;
        } else {
            $digest = "";
        }
        $digest_counts{$digest}++;
        $digest_files{$digest} //= [];
        push @{$digest_files{$digest}}, $f;
        $file_digests{$f} = $digest;
    }

    my %file_counts; # key = file name, value = num of files having file content
    for my $f (@$files) {
        next unless defined $file_sizes{$f};
        if (!defined($file_digests{$f})) {
            $file_counts{$f} = 1;
        } else {
            $file_counts{$f} = $digest_counts{ $file_digests{$f} };
        }
    }

    if ($count) {
        return [200, "OK", \%file_counts];
    } else {
        #$log->trace("report_duplicate=$report_duplicate");
        my @files;
        for (sort keys %file_counts) {
            if ($file_counts{$_} == 1) {
                #$log->trace("unique file `$_`");
                push @files, $_ if $report_unique;
            } else {
                #$log->trace("duplicate file `$_`");
                if ($report_duplicate == 1) {
                    push @files, $_;
                } elsif ($report_duplicate == 2) {
                    my $digest = $file_digests{$_};
                    push @files, $_ if $_ eq $digest_files{$digest}[0];
                }
            }
        }
        return [200, "OK", \@files];
    }
}

1;
# ABSTRACT: Report or omit duplicate file contents

__END__

=pod

=encoding UTF-8

=head1 NAME

App::UniqFiles - Report or omit duplicate file contents

=head1 VERSION

This document describes version 0.09 of App::UniqFiles (from Perl distribution App-UniqFiles), released on 2015-01-03.

=head1 SYNOPSIS

 # See uniq-files script

=head1 FUNCTIONS


=head2 uniq_files(%args) -> [status, msg, result, meta]

Report or omit duplicate file contents.

Examples:

 uniq_files();


List all files which do no have duplicate contents.


 uniq_files();


List all files which have duplicate contents.


 uniq_files();


List number of occurences of contents for each file.


Given a list of filenames, will check each file size and content for duplicate
content. Interface is a bit like the C<uniq> Unix command-line program.

Arguments ('*' denotes required arguments):

=over 4

=item * B<check_content> => I<bool> (default: 1)

Whether to check file content .

If set to 0, uniqueness will be determined solely from file size. This can be
quicker but might generate a false positive when two files of the same size are
deemed as duplicate even though their content are different.

=item * B<count> => I<bool> (default: 0)

Whether to return each file content's number of occurence.

1 means the file content is only encountered once (unique), 2 means there is one
duplicate, and so on.

=item * B<files>* => I<array[str]>

=item * B<report_duplicate> => I<int> (default: 2)

Whether to return duplicate items.

Can be set to either 0, 1, 2.

If set to 2 (the default), will only return the first of duplicate items. For
example: file1 contains text 'a', file2 'b', file3 'a'. Only file1 will be
returned because file2 is unique and file3 contains 'a' (already represented by
file1).

If set to 1, will return all the the duplicate items. From the above example:
file1 and file3 will be returned.

If set to 0, duplicate items will not be returned.

=item * B<report_unique> => I<bool> (default: 1)

Whether to return unique items.

=back

Returns an enveloped result (an array).

First element (status) is an integer containing HTTP status code
(200 means OK, 4xx caller error, 5xx function error). Second element
(msg) is a string containing error message, or 'OK' if status is
200. Third element (result) is optional, the actual result. Fourth
element (meta) is called result metadata and is optional, a hash
that contains extra information.

Return value:  (any)
=head1 NOTES

Warning: cannot properly handle symlinks or special files (socket, pipe,
device), so don't feed them.

=head1 HOMEPAGE

Please visit the project's homepage at L<https://metacpan.org/release/App-UniqFiles>.

=head1 SOURCE

Source repository is at L<https://github.com/sharyanto/perl-App-UniqFiles>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://rt.cpan.org/Public/Dist/Display.html?Name=App-UniqFiles>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

perlancar <perlancar@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by perlancar@cpan.org.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
